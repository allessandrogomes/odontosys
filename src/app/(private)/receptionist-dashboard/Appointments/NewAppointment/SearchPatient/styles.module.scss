.container {
    width: 500px;
    height: 300px;
    display: flex;
    justify-content: center;
    position: relative;

    .feedbackMsg {
        position: absolute;
        bottom: 50px;
    }
}