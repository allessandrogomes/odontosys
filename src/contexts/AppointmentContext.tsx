import { createContext, useReducer, useContext, ReactNode } from "react"

type Appointment = {
    patientName: string | null
    patientId: number | null
    dentistName: string | null
    dentistId: number | null
    scheduledAt: string | null
    endsAt: string | null
    durationMinutes: number | null
    procedure: string | null
    step: number
}

type State = Appointment

type Action =
    | { type: "SET_PATIENT"; payload: { id: number | null; name: string | null } }
    | { type: "SET_DENTIST"; payload: { id: number | null; name: string | null } }
    | { type: "SET_SCHEDULE"; payload: { scheduledAt: string | null; endsAt: string | null } }
    | { type: "SET_PROCEDURE"; payload: { procedure: string | null; durationMinutes: number | null }  }
    | { type: "SET_STEP"; payload: number }
    | { type: "RESET_APPOINTMENT" }

const initialState: State = {
    patientName: null,
    patientId: null,
    dentistName: null,
    dentistId: null,
    scheduledAt: null,
    endsAt: null,
    durationMinutes: null,
    procedure: null,
    step: 1
}

function reducer(state: State, action: Action): State {
    switch (action.type) {
        case "SET_PATIENT":
            return {
                ...state,
                patientId: action.payload.id,
                patientName: action.payload.name,
            }
        case "SET_DENTIST":
            return {
                ...state,
                dentistId: action.payload.id,
                dentistName: action.payload.name,
            }
        case "SET_SCHEDULE":
            return {
                ...state,
                scheduledAt: action.payload.scheduledAt,
                endsAt: action.payload.endsAt,
            }
        case "SET_PROCEDURE":
            return {
                ...state,
                procedure: action.payload.procedure,
                durationMinutes: action.payload.durationMinutes,
            }
        case "SET_STEP":
            return {
                ...state,
                step: action.payload
            }
        case "RESET_APPOINTMENT":
            return initialState
        default:
            return state
    }
}

const AppointmentContext = createContext<{
    state: State
    dispatch: React.Dispatch<Action>
} | undefined>(undefined)

export function AppointmentProvider({ children }: { children: ReactNode }) {
    const [state, dispatch] = useReducer(reducer, initialState)

    return (
        <AppointmentContext.Provider value={{ state, dispatch }}>
            {children}
        </AppointmentContext.Provider>
    )
}

export function useAppointmentContext() {
    const context = useContext(AppointmentContext)
    if (!context) {
        throw new Error("useAppointment must be used within an AppointmentProvider")
    }
    return context
}