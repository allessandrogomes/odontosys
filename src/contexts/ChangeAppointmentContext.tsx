import { createContext, useReducer, useContext, ReactNode } from "react"

type ChangeAppointmentState = {
    appointments: IAppointment[] | [] | null
    selectedAppointment: IAppointment | null
    step: number
}

type State = ChangeAppointmentState

type Action =
    | { type: "SET_APPOINTMENTS"; payload: IAppointment[] | [] | null }
    | { type: "SET_SELECTED_APPOINTMENT"; payload: IAppointment | null }
    | { type: "SET_STEP"; payload: number }

const initialState: State = {
    appointments: null,
    selectedAppointment: null,
    step: 1
}

function reducer(state: State, action: Action): State {
    switch (action.type) {
        case "SET_APPOINTMENTS":
            return {
                ...state,
                appointments: action.payload
            }
        case "SET_SELECTED_APPOINTMENT":
            return {
                ...state,
                selectedAppointment: action.payload
            }
        case "SET_STEP":
            return {
                ...state,
                step: action.payload
            }
        default:
            return state
    }
}

const ChangeAppointmentContext = createContext<{
    state: State
    dispatch: React.Dispatch<Action>
} | undefined>(undefined)

export function ChangeAppointmentProvider({ children }: { children: ReactNode }) {
    const [state, dispatch] = useReducer(reducer, initialState)

    return (
        <ChangeAppointmentContext.Provider value={{ state, dispatch }}>
            {children}
        </ChangeAppointmentContext.Provider>
    )
}

export function useChangeAppointmentContext() {
    const context = useContext(ChangeAppointmentContext)
    if (!context) {
        throw new Error("useChangeAppointmentContext must be used within and ChangeAppointmentProvider")
    }
    return context
}