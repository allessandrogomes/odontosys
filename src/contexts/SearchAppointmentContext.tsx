import { createContext, useReducer, useContext, ReactNode } from "react"

type SearchAppointmentState = {
    appointments: IAppointment[] | [] | null
    selectedAppointment: IAppointment | null
    step: number
}

type State = SearchAppointmentState

type Action =
    | { type: "SET_APPOINTMENTS"; payload: IAppointment[] | [] | null }
    | { type: "SET_SELECTED_APPOINTMENT"; payload: IAppointment | null }
    | { type: "SET_STEP"; payload: number }

const initialState: State = {
    appointments: null,
    selectedAppointment: null,
    step: 1
}

function reducer(state: State, action: Action): State {
    switch (action.type) {
        case "SET_APPOINTMENTS":
            return {
                ...state,
                appointments: action.payload
            }
        case "SET_SELECTED_APPOINTMENT":
            return {
                ...state,
                selectedAppointment: action.payload
            }
        case "SET_STEP":
            return {
                ...state,
                step: action.payload
            }
        default:
            return state
    }
}

const SearchAppointmentContext = createContext<{
    state: State
    dispatch: React.Dispatch<Action>
} | undefined>(undefined)

export function SearchAppointmentProvider({ children }: { children: ReactNode }) {
    const [state, dispatch] = useReducer(reducer, initialState)

    return (
        <SearchAppointmentContext.Provider value={{ state, dispatch }}>
            {children}
        </SearchAppointmentContext.Provider>
    )
}

export function useSearchAppointmentContext() {
    const context = useContext(SearchAppointmentContext)
    if (!context) {
        throw new Error("useSearchAppointment must be used within an SearchAppointmentProvider")
    }
    return context
}