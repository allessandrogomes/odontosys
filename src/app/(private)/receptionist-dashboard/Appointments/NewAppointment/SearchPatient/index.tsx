import FeedbackMessage from "@/components/ui/FeedbackMessage"
import { useState } from "react"
import { useAppointmentContext } from "@/contexts/AppointmentContext"
import { Info } from "lucide-react"
import PatientDetailsForm from "@/components/forms/PatientDetailsForm"
import styles from "./styles.module.scss"

type PatientResult =
    | { type: "patient"; patient: IPatient }
    | { type: "appointments"; appointments: IAppointment[] }
    | { type: "message"; message: string }

export default function SearchPatient() {
    const [feedbackMessage, setFeedbackMessage] = useState<string | null>(null)
    const { dispatch } = useAppointmentContext()

    // Função para lidar com o resultado da busca do paciente
    // Atualiza o estado do contexto com os dados do paciente ou exibe uma mensagem de feedback
    function handleResult(result: PatientResult) {
        switch (result.type) {
            case "patient":
                dispatch({ type: "SET_PATIENT", payload: { id: result.patient.id, name: result.patient.name } })
                setFeedbackMessage(null)

                dispatch({ type: "SET_STEP", payload: 2 })
                break
            
            case "message":
                setFeedbackMessage(result.message || "Erro ao buscar paciente")
                break

            // Não deveria ocorrer, mas garante que o tipo está coberto
            default:
                setFeedbackMessage("Erro inesperado")
                break
        }
    }

    return (
        <div className={styles.container}>
            <PatientDetailsForm onResult={handleResult} callType="patient" />
            {feedbackMessage && <FeedbackMessage className={styles.feedbackMsg} icon={<Info />} message={feedbackMessage} />}
        </div>
    )
}