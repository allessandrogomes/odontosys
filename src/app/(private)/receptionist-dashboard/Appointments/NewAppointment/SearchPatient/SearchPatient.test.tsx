import { useAppointmentContext } from "@/contexts/NewAppointmentContext"
import { render } from "@testing-library/react"
import React from "react"
import SearchPatient from "."

// Mock do contexto
jest.mock("@/contexts/NewAppointmentContext", () => ({
    useAppointmentContext: jest.fn()
}))

// Mock dos subcomponentes para evitar renderizaç~ões complexas
jest.mock("@/components/forms/PatientDetailsForm", () => ({
    __esModule: true,
    default: jest.fn(() => <div>Mocked PatientDetailsForm</div>)
}))

jest.mock("@/components/ui/FeedbackMessage", () => ({
    __esModule: true,
    default: jest.fn(({ message }) => <div>Mocked FeedbackMessage: {message}</div>)
}))

describe("SearchPatient - Renderização", () => {
    const dispatchMock = jest.fn()

    beforeEach(() => {
        (useAppointmentContext as jest.Mock).mockReturnValue({
            dispatch: dispatchMock
        })
        jest.clearAllMocks()
    })

    it ("deve renderizar corretamente sem feedbackMessage", () => {
        const { container } = render(<SearchPatient />)
        expect(container).toMatchSnapshot()
    })

    it ("deve renderizar corretamente quando há feedbackMessage", () => {
        // Para simular o feedbackMessage, podemos "forçar" o estado inicial
        const useStateSpy = jest.spyOn(React, "useState")
        useStateSpy.mockImplementation(() => ["Mensagem de teste", jest.fn()])

        const { container } = render(<SearchPatient />)
        expect(container).toMatchSnapshot()

        useStateSpy.mockRestore()
    })
})