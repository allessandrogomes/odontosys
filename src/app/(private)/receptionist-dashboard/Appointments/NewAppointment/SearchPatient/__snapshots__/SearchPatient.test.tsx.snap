// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`SearchPatient - Renderização deve renderizar corretamente quando há feedbackMessage 1`] = `
<div>
  <div
    class="container"
  >
    <div>
      Mocked PatientDetailsForm
    </div>
    <div>
      Mocked FeedbackMessage: 
      Mensagem de teste
    </div>
  </div>
</div>
`;

exports[`SearchPatient - Renderização deve renderizar corretamente sem feedbackMessage 1`] = `
<div>
  <div
    class="container"
  >
    <div>
      Mocked PatientDetailsForm
    </div>
  </div>
</div>
`;
